### PR TITLE
additional checks for 2.0: leader-follower, statefulSet, cloud-integration configs

### PR DESCRIPTION
**Pull request by JesseGoodier**
## What does this PR change?
Add additional checks for 2.0:  leader-follower, statefulSet, cloud-integration configs

## Does this PR rely on any other PRs?
No

## How does this PR impact users? (This is the kind of thing that goes in release notes!)
Additional 2.0 upgrade checks for users that have leader-follower or statefulSet deployments enabled.
Added cloud-integration check for mutually exclusive methods of providing that configuration.

## What risks are associated with merging this PR? What is required to fully test this PR?
None

## How was this PR tested?
Helm upgrade with and without the relevant values. Output of those tests below.

```
helm upgrade kubecost-cloudcost ./cost-analyzer 
Alias tip: hup kubecost-cloudcost ./cost-analyzer
Release "kubecost-cloudcost" has been upgraded. Happy Helming!
NAME: kubecost-cloudcost
LAST DEPLOYED: Thu Feb 15 14:41:33 2024
NAMESPACE: kubecost-cloudcost
STATUS: deployed
REVISION: 63
NOTES:
--------------------------------------------------
Kubecost 2.0.0 has been successfully installed.

Welcome to Kubecost 2.0!

Kubecost 2.0 is a major upgrade from previous versions and includes major new features including a brand new API Backend. Please review the following documentation to ensure a smooth transition: https://docs.kubecost.com/install-and-configure/install/kubecostv2

For the full list of enhancements, please see our release notes: https://github.com/kubecost/cost-analyzer-helm-chart/releases/tag/v2.0.0

When pods are Ready, you can enable port-forwarding with the following command:

    kubectl port-forward --namespace kubecost-cloudcost deployment/kubecost-cloudcost-cost-analyzer 9090

Then, navigate to http://localhost:9090 in a web browser.

Please allow 25 minutes for Kubecost to gather metrics. A progress indicator will appear at the top of the UI.

Having installation issues? View our Troubleshooting Guide at http://docs.kubecost.com/troubleshoot-install

(qa-gcp1/kubecost-cloudcost) ubuntu@aerog in /home/ubuntu/git/cost-analyzer-helm-chart on v2-upgrade-checks
>helm template kubecost cost-analyzer --set kubecostProductConfigs.bigQueryBillingDataDataset=test
Error: execution error at (cost-analyzer/templates/NOTES.txt:6:4): 
Kubecost 2.x requires a change to the method that cloud-provider billing integrations are configured.
Please use this output to create a cloud-integration.json config. See:
<https://docs.kubecost.com/install-and-configure/install/cloud-integration#adding-a-cloud-integration>
for more information

  {
    "gcp":
      {
        [
          {
              "bigQueryBillingDataDataset": "test",
              "bigQueryBillingDataProject": "<no value>",
              "bigQueryBillingDataTable": "<no value>",
              "projectID": "<no value>"
          }
        ]
      }
  }

Use --debug flag to render out invalid YAML

(qa-gcp1/kubecost-cloudcost) ubuntu@aerog in /home/ubuntu/git/cost-analyzer-helm-chart on v2-upgrade-checks
>helm template kubecost cost-analyzer --set kubecostProductConfigs.azureStorageContainer=test     
Error: execution error at (cost-analyzer/templates/NOTES.txt:7:4): 

Kubecost 2.x requires a change to the method that cloud-provider billing integrations are configured.
Please use this output to create a cloud-integration.json config. See:
<https://docs.kubecost.com/install-and-configure/install/cloud-integration#adding-a-cloud-integration>
for more information
  {
    "azure":
      [
        {
            "azureStorageContainer": "test",
            "azureSubscriptionID": "<no value>",
            "azureStorageAccount": "<no value>",
            "azureStorageAccessKey": "<no value>",
            "azureContainerPath": "<no value>",
            "azureCloud": "<no value>"
        }
      ]
  }

Use --debug flag to render out invalid YAML
```

### CODE REVIEW

File: cost-analyzer/templates/_helpers.tpl [hunk 1]
```diff
@@ -91,6 +91,14 @@ Kubecost 2.0 preconditions
     {{- fail "Kubecost no longer includes PodSecurityPolicy by default. Please take steps to preserve your existing PSPs before attempting the installation/upgrade again with the podSecurityPolicy values removed." }}
   {{- end }}
 
+  {{- if ((.Values.kubecostDeployment).leaderFollower).enabled -}}
+    {{- fail "\nIn Kubecost 2.0, Leader Follower architecture has been deprecated, please reach out to support to discuss upgrade paths." -}}
```
Comment (chipzoller) [inline cost-analyzer/templates/_helpers.tpl]: deprecated ≠ removed. Is it removed (not available with this release) or is it deprecated (available now but being removed in the future)?

File: cost-analyzer/templates/_helpers.tpl [hunk 2]
```diff
@@ -91,6 +91,14 @@ Kubecost 2.0 preconditions
...
+  {{- end -}}
+
+  {{- if ((.Values.kubecostDeployment).statefulSet).enabled -}}
+    {{- fail "\nIn Kubecost 2.0, kubecostDeployment does not support running as a statefulSet. This architecture has been deprecated, please reach out to support to discuss upgrade paths." -}}
```
Comment (chipzoller) [inline cost-analyzer/templates/_helpers.tpl]: Similar comment.

File: cost-analyzer/templates/_helpers.tpl [hunk 3]
```diff
@@ -149,7 +160,7 @@ Verify the cloud integration secret exists with the expected key when cloud inte
 {{-  if .Capabilities.APIVersions.Has "v1/Secret" }}
   {{- $secret := lookup "v1" "Secret" .Release.Namespace .Values.kubecostProductConfigs.cloudIntegrationSecret }}
   {{- if or (not $secret) (not (index $secret.data "cloud-integration.json")) }}
-    {{- fail (printf "The cloud integration secret '%s' does not exist or does not contain the expected key 'cloud-integration.json'" .Values.kubecostProductConfigs.cloudIntegrationSecret) }}
+    {{- fail (printf "The cloud integration secret '%s' does not exist or does not contain the expected key 'cloud-integration.json'\nIf you are using `--dry-run`, please add `--dry-run=server`" .Values.kubecostProductConfigs.cloudIntegrationSecret) }}
```
Comment (chipzoller) [inline cost-analyzer/templates/_helpers.tpl]:
```suggestion
    {{- fail (printf "The cloud integration secret '%s' does not exist or does not contain the expected key 'cloud-integration.json'\nIf you are using `--dry-run`, please add `--dry-run=server`. This requires Helm 3.13+." .Values.kubecostProductConfigs.cloudIntegrationSecret) }}
```

File: cost-analyzer/templates/_helpers.tpl [hunk 4]
```diff
@@ -118,10 +126,13 @@ will result in failure. Users are asked to select one of the two presently-avail
 */}}
 {{- define "cloudIntegrationSourceCheck" -}}
   {{- if and (.Values.kubecostProductConfigs).cloudIntegrationSecret (.Values.kubecostProductConfigs).cloudIntegrationJSON -}}
-    {{- fail "\ncloudIntegrationSecret and cloudIntegrationJSON are mutually exclusive. Please specify only one." -}}
+    {{- fail "\nkubecostProductConfigs.cloudIntegrationSecret and kubecostProductConfigs.cloudIntegrationJSON are mutually exclusive. Please specify only one." -}}
+  {{- end -}}
+  {{- if and (.Values.kubecostProductConfigs).cloudIntegrationSecret ((.Values.kubecostProductConfigs).athenaProjectID) }}
+    {{- fail "\nkubecostProductConfigs.cloudIntegrationSecret and kubecostProductConfigs.athena* values are mutually exclusive. Please specifiy only one." -}}
```
Comment (chipzoller) [inline cost-analyzer/templates/_helpers.tpl:142]: Don't these checks also need to take into consideration if users have defined either Azure and/or GCP values under the `kubecostProductConfigs` map?

File: cost-analyzer/templates/_helpers.tpl [hunk 5]
```diff
@@ -118,10 +126,13 @@ will result in failure. Users are asked to select one of the two presently-avail
...
   {{- end -}}
-{{- if and (.Values.kubecostProductConfigs).cloudIntegrationSecret ((.Values.kubecostProductConfigs).athenaProjectID) }}
-    {{- fail "\nUsing a cloud-integration secret and kubecostProductConfigs.athena* values are mutually exclusive. Please specifiy only one." -}}
+  {{- if and (.Values.kubecostProductConfigs).cloudIntegrationJSON ((.Values.kubecostProductConfigs).athenaProjectID) }}
```
Comment (chipzoller) [inline cost-analyzer/templates/_helpers.tpl]:
```suggestion
{{- if and (.Values.kubecostProductConfigs).cloudIntegrationJSON ((.Values.kubecostProductConfigs).athenaProjectID) }}
```

File: cost-analyzer/templates/_helpers.tpl [hunk 6]
```diff
@@ -1108,4 +1122,43 @@ SSO enabled flag for nginx configmap
...
+        ]
+      }
+  }
+{{- if ((.Values.kubecostProductConfigs).bigQueryBillingDataDataset) }}
+{{- fail "\nKubecost 2.0 requires a change to the method that cloud-provider billing integrations are configured.\nPlease use this output to create a cloud-integration.json config. See:\n<https://docs.kubecost.com/install-and-configure/install/cloud-integration#adding-a-cloud-integration>\nfor more information"}}
+{{- end }}
+{{- end }}
+
+
+{{- define "azureCloudIntegrationCheck" }}
```
Comment (jessegoodier) [inline cost-analyzer/templates/_helpers.tpl]: @chipzoller is there a way to fail and print this out? See the fail note for this and the gcpCloudIntegrationCheck

Comment (chipzoller) [reply cost-analyzer/templates/_helpers.tpl]: Yes, you can just define a var similar to `cloudIntegrationJSON` which is a block scalar and print this out in the failure message referencing your var.

I'll also note that the current schema for that JSON document is missing `azureContainerPath` and `azureCloud` keys. Seems like they should be in `cloudIntegrationJSON`, right?

**Comment by chipzoller** [comment]
Conflicts

**Comment by jessegoodier** [comment]
> Conflicts
fixed

**Review by chipzoller — APPROVED** [review]

**Pull request merged**